lib/matplotlib/_constrained_layout.py

```diff
@@ -106,42 +108,135 @@ def do_constrained_layout(fig, renderer, h_pad, w_pad,
 
         # make margins for all the axes and subfigures in the
         # figure.  Add margins for colorbars...
-        _make_layout_margins(fig, renderer, h_pad=h_pad, w_pad=w_pad,
-                             hspace=hspace, wspace=wspace)
-        _make_margin_suptitles(fig, renderer, h_pad=h_pad, w_pad=w_pad)
+        _make_layout_margins(_layoutgrids, fig, renderer, h_pad=h_pad,
+                             w_pad=w_pad, hspace=hspace, wspace=wspace)
+        _make_margin_suptitles(_layoutgrids, fig, renderer, h_pad=h_pad,
+                               w_pad=w_pad)
 
         # if a layout is such that a columns (or rows) margin has no
         # constraints, we need to make all such instances in the grid
         # match in margin size.
-        _match_submerged_margins(fig)
+        _match_submerged_margins(_layoutgrids, fig)
 
         # update all the variables in the layout.
-        fig._layoutgrid.update_variables()
+        _layoutgrids[fig].update_variables()
 
-        if _check_no_collapsed_axes(fig):
-            _reposition_axes(fig, renderer, h_pad=h_pad, w_pad=w_pad,
-                             hspace=hspace, wspace=wspace)
+        if _check_no_collapsed_axes(_layoutgrids, fig):
+            _reposition_axes(_layoutgrids, fig, renderer, h_pad=h_pad,
+                             w_pad=w_pad, hspace=hspace, wspace=wspace)
         else:
             _api.warn_external('constrained_layout not applied because '
                                'axes sizes collapsed to zero.  Try making '
                                'figure larger or axes decorations smaller.')
-        _reset_margins(fig)
+        _reset_margins(_layoutgrids, fig)
+    return _layoutgrids
+
+
+def _make_layoutgrids(fig, _layoutgrids):
+    """
+    Make the layoutgrid tree.
+
+    (Sub)Figures get a layoutgrid so we can have figure margins.
+
+    Gridspecs that are attached to axes get a layoutgrid so axes
+    can have margins.
+    """
+
+    if _layoutgrids is None:
+        _layoutgrids = dict()
+        _layoutgrids['hasgrids'] = False
+    if not hasattr(fig, '_parent'):
+        # top figure
+        _layoutgrids[fig] = layoutgrid.LayoutGrid(parent=None, name='figlb')
+    else:
+        # subfigure
+        gs = fig._subplotspec.get_gridspec()
+        # its possible the gridspec containing this subfigure hasn't
+        # been added to the tree yet:
+        _layoutgrids = _make_layoutgrids_gs(_layoutgrids, gs)
+        # add the layoutgrid for the subfigure:
+        parentlb = _layoutgrids[gs]
+        _layoutgrids[fig] = layoutgrid.LayoutGrid(
+            parent=parentlb,
+            name=(parentlb.name + '.' + 'panellb' +
+                    layoutgrid.seq_id()),
```

Comment:
Perhaps you can have a `seq_id(<prefix>)` function to avoid all these string manipulations everywhere.
(possibly in a separate PR)

Reply:
Sure, I just moved it into LayoutGrid.  The names were just for debug purposes to keep track of what each grid was supposed to be containing.